[WIP] solc option for selecting yul optimisations (broken draft PR)

cameel opened the pull request:
Status:
- [ ] Not fully tested yet, probably has bugs (but it compiles).
- [ ] Validations are missing.

Depends on #8424.

cameel committed:
OptimiserSuite: Define a constant for the maximum number of repeats of the optimisation string

cameel committed:
OptimiserSuite: Extract the code for repeating a sequence into runSequenceUntilStable()

cameel committed:
OptimiserSuite: Add a variant of runSequence() that works with a string of abbreviations

cameel committed:
OptimiserSuite: Replace full step names in run() with abbreviations

cameel committed:
OptimiserSuite: Allow passing a custom optimisation step sequence to run()

cameel committed:
OptimiserSettings: Add yulOptimiserSteps setting and pass it down to OptimiserSuite

cameel committed:
Include yulOptimisationSteps when serializing/deserializing yulDetails to/from JSON

cameel committed:
solc: Add --yul-optimizations command-line option

cameel commented:
@chriseth I need some feedback:
1. How documented should the option be? I.e. should the command-line help tell the allowed syntax and what the letters mean?
2. The command-line parser does not seem to be covered by tests other than `test/cmdlineTests/` but those seem to me more like end-to-end tests through the command-line interface. What tests would you suggest here (if any)?
3. I considered removing the current hard-coded string from `OptimiserSuite::run()` and adding it as a default value for `OptimiserSettings.yulOptimiserSteps` instead. I like the idea but I also see some serious downsides - it would be harder to understand out of context; you could not omit it when calling `run()` internally; we would have extra `fgo` steps added to it even though it already contains them. What's your opinion?

chriseth commented:
Command-line help should be rather brief, I think we cannot list all the steps there. End-to-End-tests through the commandline interface are fine.

A default for `yulOtimiserSteps` sounds like the right way to go.

The functionality we have to preserve is that you get the same result if you re-run the compilation with the `yulOptimiserSteps` provided in the metadata output. This means we should either always prefix `fgo` regardless of the value of `yulOptimiserSteps` or at least check if it starts with `fgo`.

cameel committed:
OptimiserSuite: Rewrite runSequence() to be easier to understand

- Until now the whole function was a state machine and you had to read it all to understand how it works. Now it's divided into smaller, independent chunks of code

cameel commented:
I'm closing this draft because I messed up the branches here and github does not allow renaming a branch in a PR. I'm going to submit the final code in a new PR.

This draft uses the branch with the same name as in #8424 but pushed to a different remote (`imapp/solidity` vs `ethereum/solidity`). My original intention was to create a copy of the branch from #8424 in `ethereum/solidity` and use it as a base here to prevent changes from that PR from showing up in the diff. I did push the branch but looks like I must then have absent-minedly used it  as a target rather than as a base.
